sql/core/src/main/scala/org/apache/spark/sql/Window.scala

```diff
@@ -0,0 +1,222 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import scala.language.implicitConversions
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.catalyst.expressions._
+
+
+sealed private[sql] class Frame(private[sql] var boundary: FrameBoundary = null)
+
+/**
+ * :: Experimental ::
+ * An utility to specify the Window Frame Range.
+ */
+object Frame {
+  val currentRow: Frame = new Frame(CurrentRow)
+  val unbounded: Frame = new Frame()
+  def preceding(n: Int): Frame = if (n == 0) {
+    new Frame(CurrentRow)
+  } else {
+    new Frame(ValuePreceding(n))
+  }
+
+  def following(n: Int): Frame = if (n == 0) {
+    new Frame(CurrentRow)
+  } else {
+    new Frame(ValueFollowing(n))
+  }
+}
+
+/**
+ * :: Experimental ::
+ * A Window object with everything unset. But can build new Window object
+ * based on it.
+ */
+@Experimental
+object Window extends Window()
```

Comment:
i think about it more -- this is actually problematic for java because none partitionBy, orderBy won't become static methods due to conflicts with the Window class. Here's the fix.

Move "class Window" into "object Window", and rename it to WindowSpec.